sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,345 @@
+		return "auth"
+	case PluginTypeDatabase:
+		return "database"
+	case PluginTypeSecrets:
+		return "secret"
+	default:
+		return "unsupported"
+	}
+}
+
+// DriverOptions are a collection of options each step driver should
+// support
+type DriverOptions struct {
+	// MountPathPrefix is an optional prefix to use when mounting the plugin. If
+	// omitted the mount path will default to the PluginName with a random suffix.
+	MountPathPrefix string
+
+	// Name is used to register the plugin. This can be arbitray but should be a
+	// reasonable value. For an example, if the plugin in test is a secret backend
+	// that generates UUIDs with the name "vault-plugin-secrets-uuid", then "uuid"
+	// or "test-uuid" would be reasonable. The name is used for lookups in the
+	// catalog. See "name" in the "Register Plugin" endpoint docs:
+	// - https://www.vaultproject.io/api-docs/system/plugins-catalog#register-plugin
+	Name string
+
+	// PluginType is the optional type of plugin. See PluginType const defined
+	// above
+	PluginType PluginType
+
+	// PluginName represents the name of the plugin that gets compiled. In the
+	// standard plugin project file layout, it represents the folder under the
+	// cmd/ folder. In the below example UUID project, the PluginName would be
+	// "uuid":
+	//
+	// vault-plugin-secrets-uuid/
+	// - backend.go
+	// - cmd/
+	// ----uuid/
+	// ------main.go
+	// - path_generate.go
+	//
+	PluginName string
+}
+
+// Step represents a single step of a test Case
+type Step struct {
+	// Operation defines what action is being taken in this step; write, read,
+	// delete, et. al.
+	Operation StepOperation
+
+	// Path is the localized request path. The mount prefix, namespace, and
+	// optionally "auth" will be automatically added.
+	Path string
+
+	// Arguments to pass in the request. These arguments represent payloads sent
+	// to the API.
+	Data map[string]interface{}
+
+	// Check is a function that is called after this step is executed in order to
+	// test that the step executed successfully. If this is not set, then the next
+	// step will be called
+	Check StepCheckFunc
+
+	// ExpectError indicates if this step is expected to return an error. If the
+	// step operation returns an error and ExpectError is not set, the test will
+	// fail immediately and the StepCheckFunc will not be ran. If ExpectError is
+	// true, the StepCheckFunc will be called (if any) and include the error from
+	// the response. It is the responsibility of the StepCheckFunc to validate the
+	// error is appropriate or not, if expected.
+	ExpectError bool
+
+	// Unauthenticated will make the request unauthenticated.
+	Unauthenticated bool
+}
+
+// StepCheckFunc is the callback used for Check in Steps.
+type StepCheckFunc func(*api.Secret, error) error
+
+// Case is a collection of tests to run for a plugin
+type Case struct {
+	// Driver is used to setup the Vault instance and provide the client used to
+	// drive the tests
+	Driver StepDriver
```

Comment:
Few questions:

1. Is driver going to be only responsible for setting up a plugin against a pre-existing vault instance? Or the intention to make it responsible for deploing/tearing down a vault instance too?

2. In a case when we need to not only register a plugin with a vault instance, but also set up/tear down some other external resources. Where do we do that? In driver setup? I think it is going to be relevant for auth plugins.

Reply:
Thanks for the feedback/questions! Replies below:

1) Right now the environments are responsible for setup and teardown of the Vault instance itself, as well as compiling the plugin under test and registering it with the Vault instance. An early, very rough first version of a Docker environment (driver) can be see on the (now poorly named) [stepwise-driver-legacy][branch], specifically [here][environment]. An example test can be found in the [transit secret engine][transit].

2) Indeed auth plugins will need this in the future. For now setting up external resources such as that are beyond the initial scope of work we've set out for this framework, so it will need to be deferred. 

I hope that answers your questions, albeit maybe not sufficiently for the second one. The primary goal of this work is to effectively overhaul the existing but not often used plugin test framework to enable better externally driven testing, using an actual instance of Vault. 

[branch]: https://github.com/hashicorp/vault/tree/stepwise-driver-legacy
[environment]: https://github.com/hashicorp/vault/blob/5b8f2f71dbd1fbb4291cb90ba4048b3f817a4f0b/sdk/testing/stepwise/environments/docker/environment.go
[transit]: https://github.com/hashicorp/vault/blob/5b8f2f71dbd1fbb4291cb90ba4048b3f817a4f0b/builtin/logical/transit/stepwise_test.go